Update stripe checkout

joshfeck opened the pull request:
This will need an initial review and some further work. 

See also:
http://p2.eventespresso.com/2017/12/08/ee3-and-stripe/

Pebblo committed:
Update Stripe library to 1.18.0

[touch: 11244]

Pebblo committed:
Update the Stripe gateway output to use Checkout.

[touch: 11244]

Pebblo committed:
Use a custom Stripe publishable key if one is set in event_meta.

[touch: 11244]

Pebblo committed:
Update Stripe class to use the newer library.

[touch: 11244]

Pebblo committed:
Load the Stripe payment method if a stripeToken has been set.

[touch: 11244]

Pebblo committed:
Use a custom Stripe secret key if one is set on event_meta.
Call the Stripe API class directly.
Pull the stripeToken from $_POST.
Use the API to set the Stripe key and create a charge.
Convert charge to an array so we can use the values.
Check for any error messages and display them, otherwise display a success message.

[touch: 11244]

Pebblo committed:
Remove 'stripe.js' from the payment method as its no longer used.
(Note this is not the stripe.js you would expected it to be)

[touch: 11244]

Pebblo committed:
Remove stipe.class.php as its not longer used.

[touch: 11244]

mnelson4 commented on gateways/stripe/do_transaction.php:
```diff
@@ -12,40 +12,55 @@ function espresso_process_stripe($payment_data) {
+    
+    //Build the Stripe data array
+    $stripe_data = array(
+        'amount' => str_replace( array(',', '.'), '', number_format( $payment_data['total_cost'], 2) ),
```
I think there are currencies that use a different number of decimal places that we should worry about. Eg Japanese Yen doesn't use any decimal places, and I think this code will change a "100 Yen" payment into a "10000 Yen". Right? iirc in EE4 we just had to hardcode a couple of exceptions like that.

Pebblo replied:
Good call, I've added this in the latest changes (I just coped the function used by EE4)

However, EE3 doesn't really handle currencies that use something other than 2 decimal places and a period for decimal anyway, there are multiple locations that will blow up or just be incorrect if you try to.

mnelson4 replied:
oh good point

mnelson4 commented on gateways/stripe/do_transaction.php:
```diff
@@ -12,40 +12,55 @@ function espresso_process_stripe($payment_data) {
+        'card' => $token,
+        'description' => $payment_data["event_name"] . 
+        				"[" . date('m-d-Y', strtotime($payment_data['start_date'])) . "]" . 
+        				" >> " . $payment_data["fname"] . " " . $payment_data["lname"]
```
might it be better to use a translatable string here? eg "%1$s [%2$s] >> %3$s %4$s", so that other languages can reorder the content?

mnelson4 replied:
oh, and include a `//translators` comment too, like on https://core.trac.wordpress.org/ticket/44360

Pebblo replied:
Done

mnelson4 commented on gateways/stripe/do_transaction.php:
```diff
@@ -12,40 +12,55 @@ function espresso_process_stripe($payment_data) {
 			echo "<div id='stripe_response'>";
-			if ($response['status'] > 0) {
-				echo "<div class='stripe_status'>" . $response['msg'] . "</div>";
+			if ($charge_array['paid'] === TRUE) {
```
FYI, at least for EE4, we're using lowercase booleans

Pebblo replied:
Done

mnelson4 commented on gateways/stripe/do_transaction.php:
```diff
@@ -12,40 +12,55 @@ function espresso_process_stripe($payment_data) {
-			if (isset($response['error_msg']) && strlen(trim($response['error_msg'])) > 0) {
-				echo "<div class='stripe_error'>ERROR: " . $response['error_msg'] . "  </div>";
+			if ( !empty($charge_array['failure_code']) ) {
+				echo "<div class='stripe_error'>ERROR: " . $charge_array['failure_code'] . " - " . $charge_array['failure_message'] . "  </div>";
```
I'd be inclined to use a translatable string here, which might especially be good for RTL languages.

Pebblo replied:
Done

mnelson4 commented on gateways/stripe/stripe_vars.php:
```diff
@@ -29,94 +32,21 @@ function espresso_display_stripe($payment_data) {
-				</p>
+			<form action="<?php echo add_query_arg(array('r_id'=>$registration_id, 'attendee_id'=>$attendee_id), get_permalink($org_options['return_url'])); ?>" method="POST" class="allow-leave-page">
+			  <script
+				src="https://checkout.stripe.com/v2/checkout.js" class="stripe-button"
```
if reasonable, all scripts should be included in the plugin. See https://developer.wordpress.org/plugins/wordpress-org/detailed-plugin-guidelines/#8-plugins-may-not-send-executable-code-via-third-party-systems where it says "all non-service related JavaScript and CSS must be included locally". If not possible because their JS is dynamic or something, then we'll let it slide

joshfeck replied:
I don't think this rule applies here (this part of the plugin isn't included on the hosted version on wp.org) and EE4 Stripe's gateway uses the same script from Stripe:

https://github.com/eventespresso/eea-stripe-gateway/blob/master/forms/BillingForm.php#L175

Although the script should be enqueued (like how it's enqueued in the EE4 Stripe gateway).

mnelson4 replied:
oups! ya good point @joshfeck

Pebblo replied:
We also can not include checkout.js within the plugin itself and keep the lowest PCI requirement levels.

mnelson4 reviewed (changes requested):
Great thanks @Pebblo. A few suggestions. Also, we could consider removing the test files from the stripe PHP library as those are probably unnecessary bloat.

Pebblo committed:
Merge remote-tracking branch 'origin/master' into FET/11244/ee3-stripe-checkout

Pebblo committed:
Add espresso_stripe.js file.

Pebblo committed:
Enqueue JS files and localize stripe args.

Pebblo committed:
Translate Stripe description

Pebblo committed:
Add Stripe button,
Add hidden inputs for Stripe data,
Ass form to post data on success.

Pebblo committed:
Update stripeToken name.

Pebblo committed:
Update attendee_id used in the query string.

Pebblo committed:
Update Stripe token name to use the new name.

Pebblo committed:
Create Stripe charge,
Use the Stripe currency from the payment method settings (default to USD),
Translate the description passed with the charge.

Pebblo committed:
Remove unused files from Stripe library.

Pebblo committed:
start_date isn't available in payment_data at this point so I've just pulled it directly from the registration. This is to contain the change within Stripe rather than having to test it across more payment methods.

Pebblo committed:
Add a function to us 0 decimal places for specific currencies currencies.

Pebblo committed:
Allow for Stripe error translation

Pebblo commented:
Thanks, Mike.

I've 'borrowed' a lot of code from EE4 for the above so a lot of it might seem a little familiar :)

mnelson4 commented on gateways/stripe/do_transaction.php:
```diff
@@ -49,13 +54,21 @@ function espresso_process_stripe($payment_data) {
-				echo "<div class='stripe_error'>ERROR: " . $charge_array['failure_code'] . " - " . $charge_array['failure_message'] . "  </div>";
+				echo "<div class='stripe_error'>" . 
+					sprintf(
+			        	/* translators: 1: error code, 2: faileure message */
```
c'est quoi "faileure"? :wink:

mnelson4 commented on gateways/stripe/stripe_vars.php:
```diff
@@ -2,20 +2,46 @@
 		$home = home_url();
 	}
 	
+	$SQL = "SELECT start_date FROM " . EVENTS_ATTENDEE_TABLE . " WHERE registration_id='%s' ORDER BY id LIMIT 1";
```
I'm pretty sure you don't need to put %s in quotes. `$wpdb->prepare` will do that for you (you may actually be unquoting the quote)

mnelson4 commented on gateways/stripe/stripe_vars.php:
```diff
@@ -2,20 +2,46 @@
+        'stripe_description' => $stripe_description,
+        'stripe_currency' => !empty($stripe_settings['stripe_currency_symbol']) ? $stripe_settings['stripe_currency_symbol'] : 'USD',
+        'stripe_panel_label' => sprintf(__('Pay %1$s Now', 'event_espresso'), '{{amount}}'),
+        'card_error_message' => __('Payment Error! Please refresh the page and try again or contact support.', 'event_espresso'),
```
may as well put these in `esc_html__` too right?

mnelson4 commented on gateways/stripe/stripe_vars.php:
```diff
@@ -2,20 +2,46 @@
+        'stripe_org_image' => $org_options['default_logo_url'],
+        'stripe_description' => $stripe_description,
+        'stripe_currency' => !empty($stripe_settings['stripe_currency_symbol']) ? $stripe_settings['stripe_currency_symbol'] : 'USD',
+        'stripe_panel_label' => sprintf(__('Pay %1$s Now', 'event_espresso'), '{{amount}}'),
```
could use a translators comment

mnelson4 commented on gateways/stripe/stripe_vars.php:
```diff
@@ -32,17 +58,12 @@ function espresso_display_stripe($payment_data) {
-				data-description="<?php echo $event_name; ?>">
-			  </script>
+			<form id="ee-stripe-form" action="<?php echo add_query_arg(array('r_id'=>$registration_id, 'attendee_id'=>$attendee_id), get_permalink($org_options['return_url'])); ?>" method="POST" class="allow-leave-page">
+				<button id="ee-stripe-button-btn"><?php _e( 'Pay Now', 'event_espresso' );?></button>
```
may as well use `esc_html__`

mnelson4 commented on gateways/stripe/stripe_vars.php:
```diff
@@ -55,3 +76,40 @@ function espresso_display_stripe($payment_data) {
+ * @param string $currency Accepted currency.
+ * @return int
+ */
+function get_stripe_decimal_places($currency = '')
```
should add `ee_`... we mostly copied this function from woocommerce, so likelyhood of other code having the same function name is high.

joshfeck replied:
or prefix with `espresso_` since that's the naming convention for EE3 function names

mnelson4 reviewed (changes requested):
looks good! a couple touchup items only

Pebblo committed:
Use esc_html_e;
Prefix espresso_ to get_stripe_decimal_places();

joshfeck commented:
Can this be finished up @Pebblo ?

Pebblo committed:
Add translator comment.

Pebblo committed:
Update the Stripe PHP library.

Pebblo committed:
Switch to use the new Stripe library;
Create/retrieve a payment intent;
(Using an idempotency_key incase the user refreshes the payment page)
Switch over to using Stripe elements;
Add additional inputs for user details and address;
Use an option to display the address details;
Enqueue additional Stripe elements CSS;
Add a hidden h4 element used for error received from Stripe;

Pebblo committed:
We no longer use eeStripeToken, check for a payment intent ID instead;

Pebblo committed:
Switch over to use a payment intent object when processing the payment;
Update the registration based on the payment intent object;
If we have an payment intent ID use that to pull the objects from Stripe;

Pebblo committed:
Update espresso_stripe.js to use payment intents;

Pebblo committed:
Add option to collect billing address fields;

Pebblo committed:
Merge remote-tracking branch 'origin/master' into FET/11244/ee3-stripe-checkout

Pebblo committed:
Remove older Stripe library;

Pebblo commented:
Sorry for the delay on this.

I ended up switching this over to use Payment intents, I was going to just get the version in core updated to use checkout v2 and do another for SD2 compliance but in the end just opted to get it all done in one go.

Mike, I learnt a lot of how the above worked from breaking down what the EE4 PM did, so credit to you as it made it easy to follow and apply to EE3.

I did spot something that got me curious though, in EE4 we use `createPaymentMethod()` and then `handleCardAction()`, in this version I just use `handleCardPayment()` to do this in all one go, should I not be?

For anyone reading this, the functions above are from: https://stripe.com/docs/stripe-js/reference

joshfeck commented:
I've tried this and it works well. One issue though, with every theme I've tested so far the placeholders and values spill outside of the billing info input. For example, the OceanWP theme:

![Screen Shot 2019-10-03 at 10 31 28 AM](https://user-images.githubusercontent.com/891156/66136363-59d79a80-e5c9-11e9-9535-82b71894ca2d.png)

Pebblo commented:
@joshfeck that's odd, I don't get that with Chrome or FF even on OceanWP

Is espresso_stripe.css loading?

joshfeck commented:
I checked and espresso_stripe.css is loading, so that's not the issue, but it's a good clue.

Turns out it's Themeroller. I've had Themeroller deactivated on my testing site. After activating the Themeroller option the input looks great.

What we could probably do is find the CSS that's making the difference, and add it to the espresso_stripe.css stylesheet.

joshfeck committed:
apply CSS fix for when default styles are used (no themeroller styles present)

joshfeck commented:
The above commit fixes the issue in the screenshot I shared earlier.

mnelson4 commented on gateways/stripe/do_transaction.php:
```diff
@@ -12,42 +12,63 @@ function espresso_process_stripe($payment_data) {
+	}
+
+	//Include the Stripe API
+	require_once (dirname(__FILE__).'/stripe-php-6.43.1/init.php');
```
This could lead to fatal errors if another plugin (eg WooCommerce) already included the Stripe library. There are two main options:

1.We should probably check if the Stripe classes were already included. I don't think `require_once` will save you here, because PHP will probably say "well I haven't included THAT file, with that filepath, yet. I just included a DIFFERENT `init.php` for Stripe in a different folder!" and then blow up because the classes were already defined. So, making sure `! class_exists('Stripe\Stripe')` before requiring that file will help avoid fatal errors when those classes were already defined by another plugin...


2. BUT, if our code runs first, and then another plugin includes the Stripe classes (without checking that they weren't already defined) there will also be a fatal error. That's why EE4 changed the namespace from `Stripe` to `EE_Stripe` (actually, `EE_Stripe_6_35_1` because I thought we wanted two versions of the stripe classes, but we ended up removing the other). Somehow I got PHPStorm to do that magically. 

I'd do option 1 if we want to get this done ASAP and are ok with waiting for customer feedback; option 2 if we want to be thorough

mnelson4 replied:
I see [you went with option 1](https://github.com/eventespresso/event-espresso-legacy/pull/10/files/3492c291955a7e4cc6d04fef37905c91e1ef30fc..fdb4e4dfdf4c50a31c115370f893675c79ec1c82#diff-a07fc3a7d048f565d32d289717135e52R26). We can put some more time into doing option 2 if it gets prioritized.

mnelson4 commented on gateways/stripe/do_transaction.php:
```diff
@@ -12,42 +12,63 @@ function espresso_process_stripe($payment_data) {
 		echo "<div id='stripe_response' class='stripe_error'>Looks like something went wrong.  Please try again or notify the website administrator.</div>";
 	}
-	//add_action('action_hook_espresso_email_after_payment', 'espresso_email_after_payment');
+	add_action('action_hook_espresso_email_after_payment', 'espresso_email_after_payment');
```
any idea why this was commented out before? I'm not as familiar with this bit of EE3. It seems like it was mistakenly commnented-out before, which meant emails were NOT being sent after a successful payment, whereas now they will be. Is that right?

Pebblo replied:
Nice catch!

Tbh, I thought I'd commented that out during testing but see that isn't the case. Looking at the other gateways in EE, they seem to have that same line, but all commented out and it's on the payment_page.php file, so I'll comment this back out again.

mnelson4 commented on gateways/stripe/stripe_vars.php:
```diff
@@ -2,17 +2,58 @@
 
-	wp_register_script( 'stripe', EVENT_ESPRESSO_PLUGINFULLURL . 'gateways/stripe/stripe.js', array( 'jquery.validate.js' ), '1.0', TRUE );
-	wp_enqueue_script( 'stripe' );	
+	$stripe_description = sprintf(
```
Apparently this variable isn't being used. It should probably be sent when creating the payment intent.

Pebblo replied:
We can't set that on the PI can we? I think it needs to go on the charge itself. I missed the above when swapping over to elements, checkout takes the description when we create the checkout but elements need it on the charge.

The only description I can see on a PI is the statement_descriptor but that should be the name of the site/company.

Pebblo replied:
Nvm, ignore me! Looking at the wrong page. Shows description on the payment intents API page:
https://stripe.com/docs/api/payment_intents/create

mnelson4 commented on gateways/stripe/stripe_vars.php:
```diff
@@ -2,17 +2,58 @@
+	wp_enqueue_style('espresso_stripe_css', EVENT_ESPRESSO_PLUGINFULLURL . 'gateways/stripe/espresso_stripe.css');
+
+	//Include the Stripe API
+	require_once (dirname(__FILE__).'/stripe-php-6.43.1/init.php');
```
see previous note about including the Stripe library without first checking the class exists (or changing the namespace on the entire library)

mnelson4 commented on gateways/stripe/stripe_vars.php:
```diff
@@ -2,17 +2,58 @@
+	// Setup a Stripe object using the secret key.
+	\Stripe\Stripe::setApiKey( $stripe_settings['stripe_secret_key']);
+
+	$payment_intent_id = isset($_REQUEST['espresso_stripe_payment_intent_id']) ? $_REQUEST['espresso_stripe_payment_intent_id'] : '';
```
I'm not clear why we would already have the payment intent ID. Isn't this always setting up the form for a new payment (with no payment intent ID yet)? 
The code is probably fine, but I'd add a comment explaining this situation for whoever reads the code in 5 years so they're not confused also.

tn3rb replied:
should sanitize that ID coming from the `$_REQUEST` array if it's not being sanitized elsewhere

Pebblo replied:
So, I had that from a different route I'd tried and in the end just left it in as I didn't see how it could hurt, as its causing confusion already I'll remove it.

The value from `$_REQUEST['espresso_stripe_payment_intent_id']` is only sent to Stripe themselves to pull in the PI object and then the object is used elsewhere but I can sanitize it.

mnelson4 commented on gateways/stripe/stripe_vars.php:
```diff
@@ -2,17 +2,58 @@
+	    	'amount' => str_replace( array(',', '.'), '', number_format( $event_cost, espresso_get_stripe_decimal_places($stripe_settings['stripe_currency_symbol'])) ),
+	     	'currency' => $stripe_settings['stripe_currency_symbol'],
+		], [
+			'idempotency_key' => $registration_id,
```
If the registration ID is used to identify a unique payment intent, what happens if a payment fails and is retried? Won't Stripe see the duplicate `idempotency_key` and assume it's a repeat of the failed payment? If that's correct, I don't think we have a unique key to use for `idempotency_key`. I'm pretty sure the EE4 integration doesn't either.

Pebblo replied:
> If the registration ID is used to identify a unique payment intent, what happens if a payment fails and is retried?

Each payment creates a charge on the payment intent (PI) object, so it doesn't need a new PI each time, just the ID of the PI to assign the charge to (which obviously doesn't change).

> Won't Stripe see the duplicate idempotency_key and assume it's a repeat of the failed payment?

No, because the idempotency_key is only being used to create a payment intent itself, the charges on that PI are created/handled by `handleCardPayment` and don't use it the above. So even with the idempotency_key set the only thing that happens is you get multiple charges on the PI, like so - https://monosnap.com/file/GgUChlXkUZymx0G6ArPupPNVw92usG

That's just from me forcing 3DS failures and then trying again (without refreshing the page and using the same card details), then refreshing the page and adding the same details, this time completing the 3DS prompt.

The above just stops the user getting to the payment options and being able to refresh the page X times ending up with X number of PI's on the Stripe account that then aren't used.

mnelson4 replied:
Hmmm I think I kinda follow. It sounds like I could use that for EE4's integration too then.

mnelson4 commented on gateways/stripe/stripe_vars.php:
```diff
@@ -29,94 +70,51 @@ function espresso_display_stripe($payment_data) {
-						<label for="cvv"><?php _e('CVC Code', 'event_espresso'); ?></label>
-						<input type="text" name="csc" id="stripe_csc" autocomplete="off" class="small required" />
-					</p>
+						<h4 class="section-title"><?php _e('Billing Information', 'event_espresso') ?></h4>
```
:gun: BAM BAM BAM! Found another one of those `_e()`s eating carrots in the back yard. Them's make good soup tonight!
(May as well switch to them all to `esc_html_e()`)

Pebblo replied:
Lol, done.

mnelson4 commented on gateways/stripe/espresso_stripe.js:
```diff
@@ -0,0 +1,139 @@
+			//console.log( 'initialize', 'set_listener_for_submit_payment_button', true );
+			this.submitPaymentButton.on( 'click', function(e) {
+				e.preventDefault();
+				const billingDetails = {};
```
:police_car:  "I'm sorry sir, are you allowed to use that `const`? Let me see your build process that uses Babel.js to make it compatible with older browsers. What's that? You aren't using a build process here? Tsk tsk..."

Should probably just use `var` if we want this to be compatible with older browsers. Sorry. (FYI I'm speaking from having made this sort of mistake myself recently.)

mnelson4 commented on gateways/stripe/espresso_stripe.js:
```diff
@@ -0,0 +1,139 @@
+							billing_details: billingDetails
+						}
+					} 
+				).then( ( response ) => {
```
:police_car: "You're also not permitted to use arrow functions. Besides, you don't need it- you've already got `this` stuffed into your `stipe_instance` you greedy thief."

mnelson4 commented on gateways/stripe/espresso_stripe.js:
```diff
@@ -0,0 +1,139 @@
+		 * @function set_listener_for_submit_payment_button
+		 */
+		this.set_listener_for_submit_payment_button = function() {
+			const stipe_instance = this;
```
`stRipe_instance` :man_teacher:

Pebblo replied:
Thanks! Fixed.

mnelson4 commented on gateways/stripe/espresso_stripe.js:
```diff
@@ -0,0 +1,139 @@
+				if ( Object.keys( address ).length > 0 ) {
+					billingDetails.address = address;
+				}
+				stipe_instance.stripe.handleCardPayment( 
```
I think using `handleCardPayment` turns out to be fine because you're doing the "[Automatic confirmation](https://stripe.com/docs/payments/payment-intents/web)" approach; whereas EE4 uses "[Manual confirmation](https://stripe.com/docs/payments/payment-intents/web-manual)".
I originally thought Automatic Confirmation required an IPN to the server, but it looks like the message they send is just to the JS, which sounds perfectly acceptable.
It sounds like Automatic Confirmation has you: 
1. Server side: Create the payment intent
2. Client side: Use that, and the billing form, to charge card and store the result
3. Server side: Update the payment in our system :ok_hand: 

Manual Confirmation has you:
1. Client side: Create a Stripe payment method and store its ID
2. Server side: Create a payment intent, and use the payment method's ID to charge the card UNLESS it requires 3D secure, in which case...
3. Client side: We do 3D secure, and then store the 3D-authed payment ID
4. Server side: Attempt to complete the payment server-side

I think Automatic confirmation sounds fine provided the payment amount won't change once we show them the billing form (or we listen for when the payment amount changes, and then create a new payment intent server-side, and then update the client-side payment intent ID variable.) I don't think EE3 needs to worry about the payment amount changing once the billing form is shown right? (Whereas for EE4 it was more of an issue). So Automatic Confirmation (`handleCardPayment`) may be the better choice for EE3, whereas Manual Confirmation (`handleCardAction`) may still be better for EE4. :smiley:

mnelson4 reviewed (approved):
For EE4 you could choose between Stripe Checkout (legacy) or Stripe Elements, but this seems to just use Elements. I'm fine with that (it's simpler) but you guys don't worry about rioting from many users do you? (There will always be a few, of course.)

Also I'm not sure why `languages/event_espresso-en_US.mo` got changed (I'm also super rusty on managing translation files). Just double-checking that was on purpose.

There's some suggestions that should avoid the odd PHP fatal error and make it more compatible with older browsers.

mnelson4 reviewed (changes requested):
oups meant to request changes

joshfeck commented:
> For EE4 you could choose between Stripe Checkout (legacy) or Stripe Elements, but this seems to just use Elements. I'm fine with that (it's simpler) but you guys don't worry about rioting from many users do you?

EE3 didn't have Stripe Checkout ever, so this update doesn't change much UI-wise. It mostly brings the EE3 Stripe option from 2011 to 2019 in one go.

Pebblo committed:
Don't check if there's a payment intent ID set already;
Add description;
Setup $intent_id and $intent_client_secret vars if we have a PaymentIntent object;

Pebblo committed:
Use the intent variables;

Pebblo committed:
Use esc_html_e();

Pebblo committed:
Formatting.

Pebblo committed:
Use var in place of const;
Fix my broken r button;
Don't use arrow functions;

Pebblo committed:
Comment out the hook to fire emails again.

Pebblo committed:
Check if Stripe class exists before including another.

Pebblo committed:
Remove .mo file, this is sideloaded.

mnelson4 commented on gateways/stripe/stripe_vars.php:
```diff
@@ -134,29 +141,29 @@ function espresso_display_stripe($payment_data) {
-        default :
-            return 2;
-    }
+	if (!$currency) {
```
hmmm this is new though. At least for EE4 we're using spaces instead of tabs (see https://github.com/eventespresso/event-espresso-core/blob/master/docs/A--Best-Practices/php-standards.md#white-space and https://www.php-fig.org/psr/psr-2/).

Pebblo replied:
Thanks, Mike. I just went with tabs for consistency.

Switching between both in the same file was annoying me so I converted everything.

mnelson4 reviewed (approved):
Yeah this gets all the issues! Although I noticed a new annoying issue with regards to whitespace and coding standards. ...But I'm also realizing we use tabs everywhere else in EE3 so it's legacy code and whatever. Maybe consistency is better, so whatever you think.

joshfeck commented:
works for me

sethshoultes commented:
I am getting an error when trying to test this:

`Notice: wpdb::prepare was called incorrectly. The query does not contain the correct number of placeholders (1) for the number of arguments passed (2). Please see Debugging in WordPress for more information. (This message was added in version 4.8.3.) in /home2/web801/public_html/ee4-testing/wp-includes/functions.php on line 4778`

joshfeck commented:
Hey Seth did you checkout the FET/11244/ee3-stripe-checkout branch from GitHub?

The one place that the Stripe gateways uses wpdb prepare is lines 13-14 in stripe_vars.php:

```
$SQL = "SELECT start_date FROM " . EVENTS_ATTENDEE_TABLE . " WHERE registration_id=%s ORDER BY id LIMIT 1";
$payment_data['start_date'] = $wpdb->get_var( $wpdb->prepare( $SQL, $payment_data['registration_id']));
```

which has one placeholder and one argument.

sethshoultes commented:
@joshfeck yes, that's what's installed and activated:
![ee-legacy-stripe](https://user-images.githubusercontent.com/235315/68709451-f55a1480-0552-11ea-945e-04bff0f11107.jpg)

sethshoultes commented:
@joshfeck I think there may have been something misconfigured on the site I was testing with. I created a new site and the registration process and Stripe payment option are working now.

![image](https://user-images.githubusercontent.com/235315/68711006-f3de1b80-0555-11ea-9c66-39d9d0510497.png)

Pebblo commented:
Thanks guys, had me worried for a minute there Seth!

Merging in.
